### ISSUE TITLE
Custom events tracking did not start at first run

### ISSUE THREAD

**Issue by davidedandrea** [mapp-digital/MappIntelligence-iOS-v5#9]
We import the `5.0.6.2` version of the MappIntelligence SDK in our app.

We use `trackCustomEvent` and `trackCustomPage` methods in order to perform video / page view events tracking.

When our app is launched for the first time, it shows some system alerts relating to user permissions. Meanwhile, we initialize your SDK as illustrated in the [documentation](https://documentation.mapp.com/latest/en/ios-sdk-v5-19103956.html).

We noticed that, when the app starts for the first time, we can’t find any video events and page views tracking that we are trying to perform using the methods exposed by the SDK, neither in the library logs, nor from HTTP traffic capture tools. 
Below is an example of the library output in the console:

```
    2024-01-29 15:56:01.215098+0100 TestApp[1009:35875] [MappIntelligence Info] Autotracking is enabled: Yes
    2024-01-29 15:56:01.215181+0100 TestApp[1009:35875] [MappIntelligence Info] Batch support is enabled: No
    2024-01-29 15:56:01.215213+0100 TestApp[1009:35875] [MappIntelligence Info] Background sendout support is enabled: No
    2024-01-29 15:56:01.215243+0100 TestApp[1009:35875] [MappIntelligence Info] Number of requests in queue: 100
    2024-01-29 15:56:01.215273+0100 TestApp[1009:35875] [MappIntelligence Info] Request time interval in minutes: 0.750000
    2024-01-29 15:56:01.215301+0100 TestApp[1009:35875] [MappIntelligence Info] Log level is:  All
    2024-01-29 15:56:01.215336+0100 TestApp[1009:35875] [MappIntelligence Info] TrackIDs: (
    2024-01-29 15:56:01.215377+0100 TestApp[1009:35875] [MappIntelligence Info] Track domain is: ********
    2024-01-29 15:56:01.215401+0100 TestApp[1009:35875] [MappIntelligence Info] View Controller auto tracking is enabbled: Yes
    2024-01-29 15:56:01.264314+0100 TestApp[1009:35875] [MappIntelligence Debug] Remove requests that are older than 14 days.
    2024-01-29 15:56:49.302685+0100 TestApp[1009:35875] [MappIntelligence Debug] save current date for session detection 2024-01-29 14:56:49 +0000 with defaults 0
```

While on subsequent launches, the tracking works as expected, and logs are found in the console and by HTTP traffic capture tools. An example of successful tracking:

```
    2024-01-29 15:52:42.478701+0100 TestApp[972:33457] [MappIntelligence Info] Autotracking is enabled: Yes
    2024-01-29 15:52:42.478780+0100 TestApp[972:33457] [MappIntelligence Info] Batch support is enabled: No
    2024-01-29 15:52:42.478813+0100 TestApp[972:33457] [MappIntelligence Info] Background sendout support is enabled: No
    2024-01-29 15:52:42.478868+0100 TestApp[972:33457] [MappIntelligence Info] Number of requests in queue: 100
    2024-01-29 15:52:42.478901+0100 TestApp[972:33457] [MappIntelligence Info] Request time interval in minutes: 0.750000
    2024-01-29 15:52:42.478932+0100 TestApp[972:33457] [MappIntelligence Info] Log level is:  All
    2024-01-29 15:52:42.478970+0100 TestApp[972:33457] [MappIntelligence Info] TrackIDs: (
    2024-01-29 15:52:42.479016+0100 TestApp[972:33457] [MappIntelligence Info] Track domain is: *********
    2024-01-29 15:52:42.479042+0100 TestApp[972:33457] [MappIntelligence Info] View Controller auto tracking is enabbled: Yes
    2024-01-29 15:52:42.527874+0100 TestApp[972:33457] [MappIntelligence Debug] Remove requests that are older than 14 days.
    2024-01-29 15:52:44.344545+0100 TestApp[972:33457] [MappIntelligence Debug] save current date for session detection 2024-01-29 14:52:44 +0000 with defaults 0
    2024-01-29 15:52:45.529375+0100 TestApp[972:33765] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:45.531825+0100 TestApp[972:33765] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:45.535063+0100 TestApp[972:33765] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:45.537205+0100 TestApp[972:33765] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:45.539086+0100 TestApp[972:33765] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:45.750740+0100 TestApp[972:33759] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:45.762547+0100 TestApp[972:33759] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:46.528594+0100 TestApp[972:33770] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:46.551184+0100 TestApp[972:33770] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:52:52.278949+0100 TestApp[972:33759] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:05.046200+0100 TestApp[972:34079] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:05.050538+0100 TestApp[972:34079] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:07.904257+0100 TestApp[972:34080] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:07.948682+0100 TestApp[972:34079] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:16.110725+0100 TestApp[972:34079] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:16.114989+0100 TestApp[972:34079] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:17.097266+0100 TestApp[972:34104] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:17.101755+0100 TestApp[972:34104] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:18.105466+0100 TestApp[972:33769] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:18.110184+0100 TestApp[972:33769] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:19.102206+0100 TestApp[972:33772] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:19.108518+0100 TestApp[972:33772] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:20.106980+0100 TestApp[972:34208] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:20.111530+0100 TestApp[972:34208] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:21.103008+0100 TestApp[972:33770] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:21.107731+0100 TestApp[972:33770] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:22.097444+0100 TestApp[972:33772] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:22.102114+0100 TestApp[972:33772] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:23.103951+0100 TestApp[972:34082] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:23.108795+0100 TestApp[972:34082] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:24.098491+0100 TestApp[972:34082] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:24.106481+0100 TestApp[972:34082] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:25.099447+0100 TestApp[972:34079] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:25.103831+0100 TestApp[972:34079] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:26.104406+0100 TestApp[972:33770] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:26.109170+0100 TestApp[972:33770] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:27.100978+0100 TestApp[972:33770] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:27.106160+0100 TestApp[972:33770] [MappIntelligence Debug] request written with success: 1
    2024-01-29 15:53:27.555401+0100 TestApp[972:33457] [MappIntelligence Info] Tracking Request: **********
    2024-01-29 15:53:27.557173+0100 TestApp[972:33457] [MappIntelligence Info] Tracking Request: **********
```

In the first case (when the app starts from scratch) it seems that the tracking requests made by our app never happen.
By investigating the SDK code, we identified a use case that could lead to this behavior:

- In the `trackWithCustomEvent` method of the `MIDefaultTracker`, the SDK manages a lock/unlock mechanism to wait for a ready state before performing tracking requests.
- The `ready` flag is set to true in the `fireSignal` method (which is invoked in the init of the class and in the `updateFirstSessionWith` method).
- But at the next `trackWithCustomEvent` calls, the ready value is still false at the point of lock/unlock management, and therefore the NSCondition used remains in the wait state.
- Since our app shows system alerts in the first run, the app became inactive, calling your `willResignActive` selector,  where a `MIDefaultTracker.initHibernate` method is invoked, which sets the ready flag to false again. When it returns active, the SDK listens to the `didBecomeActive` notification, calling a selector method that does nothing.
- Consequently, subsequent event tracking remain blocked.

Is the use case we identified correct?

**Comment by MihajloMapp** [comment]
We need additional information for this task. On which IOS version is this problem reflected?

**Comment by davidedandrea** [comment]
We encountered this issue on iOS 17.x as well as on iOS 16.x

**Comment by stellatores** [comment]
This was fixed in version 5.0.7. (https://github.com/mapp-digital/MappIntelligence-iOS-v5/releases/tag/5.0.7)
